File: resources/ios/Sources/GoogleMobileAdsFunctions.swift
import Foundation
import GoogleMobileAds
import UIKit

@objc public class GoogleMobileAdsFunctions: NSObject {

    // MARK: - Initialize

    @objc public class Initialize: NSObject, BridgeFunction {
        public func execute(parameters: [String: Any]) -> [String: Any] {
            guard let appId = parameters["app_id"] as? String, !appId.isEmpty else {
                return BridgeResponse.error("app_id is required")
            }

            DispatchQueue.main.async {
                MobileAds.shared.start { _ in }
            }

            return BridgeResponse.success(["status": "initialized"])
        }
    }

    // MARK: - Banner

    @objc public class ShowBanner: NSObject, BridgeFunction {
        public func execute(parameters: [String: Any]) -> [String: Any] {
            guard let adUnitId = parameters["ad_unit_id"] as? String else {
                return BridgeResponse.error("ad_unit_id is required")
            }

            let position = parameters["position"] as? String ?? "bottom"
            let size = parameters["size"] as? String ?? "adaptive"

            DispatchQueue.main.async {
                guard let rootVC = UIApplication.shared.keyWindow?.rootViewController else { return }

                AdViewHolder.shared.bannerView?.removeFromSuperview()

                let bannerView = BannerView()
                bannerView.adUnitID = adUnitId
                bannerView.rootViewController = rootVC
                bannerView.adSize = Self.resolveAdSize(size)

                bannerView.eventDelegate = BannerEventDelegate(adUnitId: adUnitId)

                let request = Request()
                bannerView.load(request)

                let window = UIApplication.shared.keyWindow
                window?.addSubview(bannerView)

                bannerView.translatesAutoresizingMaskIntoConstraints = false
                let verticalAnchor = position == "top"
                    ? bannerView.topAnchor.constraint(equalTo: window!.safeAreaLayoutGuide.topAnchor)
                    : bannerView.bottomAnchor.constraint(equalTo: window!.safeAreaLayoutGuide.bottomAnchor)

                NSLayoutConstraint.activate([
                    bannerView.centerXAnchor.constraint(equalTo: window!.centerXAnchor),
                    verticalAnchor,
                ])

                AdViewHolder.shared.bannerView = bannerView
            }

            return BridgeResponse.success(["status": "loading"])
        }

        private static func resolveAdSize(_ size: String) -> AdSize {
            switch size {
            case "banner":            return AdSizeBanner
            case "large_banner":      return AdSizeLargeBanner
            case "medium_rectangle":  return AdSizeMediumRectangle
            default:
                let width = UIScreen.main.bounds.width
                return currentOrientationAnchoredAdaptiveBanner(withWidth: width)
            }
        }
    }

    @objc public class HideBanner: NSObject, BridgeFunction {
        public func execute(parameters: [String: Any]) -> [String: Any] {
            DispatchQueue.main.async {
                AdViewHolder.shared.bannerView?.removeFromSuperview()
                AdViewHolder.shared.bannerView = nil
            }
            return BridgeResponse.success(["status": "hidden"])
        }
    }

    // MARK: - Interstitial

    @objc public class LoadInterstitial: NSObject, BridgeFunction {
        public func execute(parameters: [String: Any]) -> [String: Any] {
            guard let adUnitId = parameters["ad_unit_id"] as? String else {
                return BridgeResponse.error("ad_unit_id is required")
            }

            DispatchQueue.main.async {
                InterstitialAd.load(
                    withAdUnitID: adUnitId,
                    request: Request()
                ) { ad, error in
                    if let error = error {
                        LaravelBridge.shared.send?(
                            "NativePHP\\GoogleMobileAds\\Events\\AdFailedToLoad",
                            [
                                "adType": "interstitial",
                                "adUnitId": adUnitId,
                                "errorCode": (error as NSError).code,
                                "errorMessage": error.localizedDescription,
                            ]
                        )
                        return
                    }

                    AdViewHolder.shared.interstitial = ad
                    LaravelBridge.shared.send?(
                        "NativePHP\\GoogleMobileAds\\Events\\AdLoaded",
                        ["adType": "interstitial", "adUnitId": adUnitId]
                    )
                }
            }

            return BridgeResponse.success(["status": "loading"])
        }
    }

    @objc public class ShowInterstitial: NSObject, BridgeFunction {
        public func execute(parameters: [String: Any]) -> [String: Any] {
            guard let ad = AdViewHolder.shared.interstitial else {
                return BridgeResponse.error("No interstitial ad loaded. Call LoadInterstitial first.")
            }

            DispatchQueue.main.async {
                guard let rootVC = UIApplication.shared.keyWindow?.rootViewController else { return }
                let delegate = FullScreenDelegate(adType: "interstitial")
                ad.fullScreenContentDelegate = delegate
                AdViewHolder.shared.interstitialDelegate = delegate
                AdViewHolder.shared.interstitial = nil
                ad.present(fromRootViewController: rootVC)
            }

            return BridgeResponse.success(["status": "showing"])
        }
    }

    // MARK: - Rewarded

    @objc public class LoadRewarded: NSObject, BridgeFunction {
        public func execute(parameters: [String: Any]) -> [String: Any] {
            guard let adUnitId = parameters["ad_unit_id"] as? String else {
                return BridgeResponse.error("ad_unit_id is required")
            }

            DispatchQueue.main.async {
                RewardedAd.load(
                    withAdUnitID: adUnitId,
                    request: Request()
                ) { ad, error in
                    if let error = error {
                        LaravelBridge.shared.send?(
                            "NativePHP\\GoogleMobileAds\\Events\\AdFailedToLoad",
                            [
                                "adType": "rewarded",
                                "adUnitId": adUnitId,
                                "errorCode": (error as NSError).code,
                                "errorMessage": error.localizedDescription,
                            ]
                        )
                        return
                    }

                    AdViewHolder.shared.rewarded = ad
                    LaravelBridge.shared.send?(
                        "NativePHP\\GoogleMobileAds\\Events\\AdLoaded",
                        ["adType": "rewarded", "adUnitId": adUnitId]
                    )
                }
            }

            return BridgeResponse.success(["status": "loading"])
        }
    }

    @objc public class ShowRewarded: NSObject, BridgeFunction {
        public func execute(parameters: [String: Any]) -> [String: Any] {
            guard let ad = AdViewHolder.shared.rewarded else {
                return BridgeResponse.error("No rewarded ad loaded. Call LoadRewarded first.")
            }

            DispatchQueue.main.async {
                guard let rootVC = UIApplication.shared.keyWindow?.rootViewController else { return }
                let delegate = FullScreenDelegate(adType: "rewarded")
                ad.fullScreenContentDelegate = delegate
                AdViewHolder.shared.rewardedDelegate = delegate
                AdViewHolder.shared.rewarded = nil

                ad.present(fromRootViewController: rootVC) {
                    let reward = ad.adReward
                    LaravelBridge.shared.send?(
                        "NativePHP\\GoogleMobileAds\\Events\\RewardEarned",
                        [
                            "rewardType": reward.type,
                            "rewardAmount": reward.amount.intValue,
                        ]
                    )
                }
            }

            return BridgeResponse.success(["status": "showing"])
        }
    }

    // MARK: - Rewarded Interstitial

    @objc public class LoadRewardedInterstitial: NSObject, BridgeFunction {
        public func execute(parameters: [String: Any]) -> [String: Any] {
            guard let adUnitId = parameters["ad_unit_id"] as? String else {
                return BridgeResponse.error("ad_unit_id is required")
            }

            DispatchQueue.main.async {
                RewardedInterstitialAd.load(
                    withAdUnitID: adUnitId,
                    request: Request()
                ) { ad, error in
                    if let error = error {
                        LaravelBridge.shared.send?(
                            "NativePHP\\GoogleMobileAds\\Events\\AdFailedToLoad",
                            [
                                "adType": "rewarded_interstitial",
                                "adUnitId": adUnitId,
                                "errorCode": (error as NSError).code,
                                "errorMessage": error.localizedDescription,
                            ]
                        )
                        return
                    }

                    AdViewHolder.shared.rewardedInterstitial = ad
                    LaravelBridge.shared.send?(
                        "NativePHP\\GoogleMobileAds\\Events\\AdLoaded",
                        ["adType": "rewarded_interstitial", "adUnitId": adUnitId]
                    )
                }
            }

            return BridgeResponse.success(["status": "loading"])
        }
    }

    @objc public class ShowRewardedInterstitial: NSObject, BridgeFunction {
        public func execute(parameters: [String: Any]) -> [String: Any] {
            guard let ad = AdViewHolder.shared.rewardedInterstitial else {
                return BridgeResponse.error("No rewarded interstitial ad loaded. Call LoadRewardedInterstitial first.")
            }

            DispatchQueue.main.async {
                guard let rootVC = UIApplication.shared.keyWindow?.rootViewController else { return }
                let delegate = FullScreenDelegate(adType: "rewarded_interstitial")
                ad.fullScreenContentDelegate = delegate
                AdViewHolder.shared.rewardedInterstitialDelegate = delegate
                AdViewHolder.shared.rewardedInterstitial = nil

                ad.present(fromRootViewController: rootVC, userDidEarnRewardHandler: {
                    let reward = ad.adReward
                    LaravelBridge.shared.send?(
                        "NativePHP\\GoogleMobileAds\\Events\\RewardEarned",
                        [
                            "rewardType": reward.type,
                            "rewardAmount": reward.amount.intValue,
                        ]
                    )
                })
            }

            return BridgeResponse.success(["status": "showing"])
        }
    }

    // MARK: - App Open

    @objc public class LoadAppOpen: NSObject, BridgeFunction {
        public func execute(parameters: [String: Any]) -> [String: Any] {
            guard let adUnitId = parameters["ad_unit_id"] as? String else {
                return BridgeResponse.error("ad_unit_id is required")
            }

            DispatchQueue.main.async {
                AppOpenAd.load(
                    withAdUnitID: adUnitId,
                    request: Request()
                ) { ad, error in
                    if let error = error {
                        LaravelBridge.shared.send?(
                            "NativePHP\\GoogleMobileAds\\Events\\AdFailedToLoad",
                            [
                                "adType": "app_open",
                                "adUnitId": adUnitId,
                                "errorCode": (error as NSError).code,
                                "errorMessage": error.localizedDescription,
                            ]
                        )
                        return
                    }

                    AdViewHolder.shared.appOpen = ad
                    LaravelBridge.shared.send?(
                        "NativePHP\\GoogleMobileAds\\Events\\AdLoaded",
                        ["adType": "app_open", "adUnitId": adUnitId]
                    )
                }
            }

            return BridgeResponse.success(["status": "loading"])
        }
    }

    @objc public class ShowAppOpen: NSObject, BridgeFunction {
        public func execute(parameters: [String: Any]) -> [String: Any] {
            guard let ad = AdViewHolder.shared.appOpen else {
                return BridgeResponse.error("No app open ad loaded. Call LoadAppOpen first.")
            }

            DispatchQueue.main.async {
                guard let rootVC = UIApplication.shared.keyWindow?.rootViewController else { return }
                let delegate = FullScreenDelegate(adType: "app_open")
                ad.fullScreenContentDelegate = delegate
                AdViewHolder.shared.appOpenDelegate = delegate
                AdViewHolder.shared.appOpen = nil
                ad.present(fromRootViewController: rootVC)
            }

            return BridgeResponse.success(["status": "showing"])
        }
    }
}

// MARK: - Helpers

private class BannerEventDelegate: NSObject, BannerViewDelegate {
    private let adUnitId: String

    init(adUnitId: String) {
        self.adUnitId = adUnitId
    }

    func bannerViewDidReceiveAd(_ bannerView: BannerView) {
        // adSize.size.height is the points height AdMob actually allocated for
        // this banner — the host app needs this to reserve exactly enough
        // space for its own layout instead of guessing a static value.
        LaravelBridge.shared.send?(
            "NativePHP\\GoogleMobileAds\\Events\\AdLoaded",
            ["adType": "banner", "adUnitId": adUnitId, "heightDp": Int(bannerView.adSize.size.height)]
        )
    }

    func bannerView(_ bannerView: BannerView, didFailToReceiveAdWithError error: Error) {
        LaravelBridge.shared.send?(
            "NativePHP\\GoogleMobileAds\\Events\\AdFailedToLoad",
            [
                "adType": "banner",
                "adUnitId": adUnitId,
                "errorCode": (error as NSError).code,
                "errorMessage": error.localizedDescription,
            ]
        )
    }

    func bannerViewWillPresentScreen(_ bannerView: BannerView) {
        LaravelBridge.shared.send?(
            "NativePHP\\GoogleMobileAds\\Events\\AdOpened",
            ["adType": "banner"]
        )
    }

    func bannerViewDidDismissScreen(_ bannerView: BannerView) {
        LaravelBridge.shared.send?(
            "NativePHP\\GoogleMobileAds\\Events\\AdClosed",
            ["adType": "banner"]
        )
    }

    func bannerViewDidRecordImpression(_ bannerView: BannerView) {
        LaravelBridge.shared.send?(
            "NativePHP\\GoogleMobileAds\\Events\\AdImpression",
            ["adType": "banner"]
        )
    }

    func bannerViewDidRecordClick(_ bannerView: BannerView) {
        LaravelBridge.shared.send?(
            "NativePHP\\GoogleMobileAds\\Events\\AdClicked",
            ["adType": "banner"]
        )
    }
}

private class FullScreenDelegate: NSObject, FullScreenContentDelegate {
    private let adType: String

    init(adType: String) {
        self.adType = adType
    }

    func adDidPresentFullScreenContent(_ ad: FullScreenPresentingAd) {
        LaravelBridge.shared.send?(
            "NativePHP\\GoogleMobileAds\\Events\\AdOpened",
            ["adType": adType]
        )
    }

    func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        LaravelBridge.shared.send?(
            "NativePHP\\GoogleMobileAds\\Events\\AdClosed",
            ["adType": adType]
        )
    }

    func ad(_ ad: FullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        LaravelBridge.shared.send?(
            "NativePHP\\GoogleMobileAds\\Events\\AdFailedToLoad",
            [
                "adType": adType,
                "adUnitId": "",
                "errorCode": (error as NSError).code,
                "errorMessage": error.localizedDescription,
            ]
        )
    }

    func adDidRecordImpression(_ ad: FullScreenPresentingAd) {
        LaravelBridge.shared.send?(
            "NativePHP\\GoogleMobileAds\\Events\\AdImpression",
            ["adType": adType]
        )
    }

    func adDidRecordClick(_ ad: FullScreenPresentingAd) {
        LaravelBridge.shared.send?(
            "NativePHP\\GoogleMobileAds\\Events\\AdClicked",
            ["adType": adType]
        )
    }
}

// MARK: - Ad Holder

private class AdViewHolder {
    static let shared = AdViewHolder()

    var bannerView: BannerView?
    var interstitial: InterstitialAd?
    var interstitialDelegate: FullScreenDelegate?
    var rewarded: RewardedAd?
    var rewardedDelegate: FullScreenDelegate?
    var rewardedInterstitial: RewardedInterstitialAd?
    var rewardedInterstitialDelegate: FullScreenDelegate?
    var appOpen: AppOpenAd?
    var appOpenDelegate: FullScreenDelegate?
}

File: resources/android/src/GoogleMobileAdsFunctions.kt
package com.nativephp.plugins.googlemobileads

import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.Gravity
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.fragment.app.FragmentActivity
import com.google.android.gms.ads.AdError
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.MobileAds
import com.google.android.gms.ads.OnUserEarnedRewardListener
import com.google.android.gms.ads.appopen.AppOpenAd
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import com.google.android.gms.ads.rewarded.RewardedAd
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback
import com.google.android.gms.ads.rewardedinterstitial.RewardedInterstitialAd
import com.google.android.gms.ads.rewardedinterstitial.RewardedInterstitialAdLoadCallback
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse
import com.nativephp.mobile.utils.NativeActionCoordinator
import org.json.JSONObject

private const val TAG = "GoogleMobileAds"
private val mainHandler = Handler(Looper.getMainLooper())

// ── Ad instance holder ────────────────────────────────────────────────────────

private object AdHolder {
    var bannerView: AdView? = null
    var interstitial: InterstitialAd? = null
    var rewarded: RewardedAd? = null
    var rewardedInterstitial: RewardedInterstitialAd? = null
    var appOpen: AppOpenAd? = null
}

// ── Event dispatch helper ─────────────────────────────────────────────────────

/**
 * Was originally hand-rolled WebView JS injection requiring `WebViewProvider`,
 * which (a) force-created a hidden WebView via the non-null `getWebView()` on
 * apps with no WebView anywhere, and (b) never reached
 * `NativeElementBridge.sendNativeEvent()`, so `#[On(...)]` listeners on a
 * native-UI-only screen never fired no matter what. `NativeActionCoordinator`
 * is the core's own dispatcher — same one Camera/Microphone use — and already
 * handles both the WebView-optional case and the native element event queue.
 */
private fun FragmentActivity.dispatchAdEvent(eventClass: String, payload: Map<String, Any>) {
    val payloadJson = JSONObject(payload).toString()
    NativeActionCoordinator.dispatchEvent(this, eventClass, payloadJson)
}

// ── Full-screen callback factory ──────────────────────────────────────────────

private fun fullScreenCallback(activity: FragmentActivity, adType: String): FullScreenContentCallback =
    object : FullScreenContentCallback() {
        override fun onAdShowedFullScreenContent() {
            activity.dispatchAdEvent(
                "NativePHP\\GoogleMobileAds\\Events\\AdOpened",
                mapOf("adType" to adType)
            )
        }

        override fun onAdDismissedFullScreenContent() {
            activity.dispatchAdEvent(
                "NativePHP\\GoogleMobileAds\\Events\\AdClosed",
                mapOf("adType" to adType)
            )
        }

        override fun onAdFailedToShowFullScreenContent(error: AdError) {
            activity.dispatchAdEvent(
                "NativePHP\\GoogleMobileAds\\Events\\AdFailedToLoad",
                mapOf("adType" to adType, "adUnitId" to "", "errorCode" to error.code, "errorMessage" to error.message)
            )
        }

        override fun onAdImpression() {
            activity.dispatchAdEvent(
                "NativePHP\\GoogleMobileAds\\Events\\AdImpression",
                mapOf("adType" to adType)
            )
        }

        override fun onAdClicked() {
            activity.dispatchAdEvent(
                "NativePHP\\GoogleMobileAds\\Events\\AdClicked",
                mapOf("adType" to adType)
            )
        }
    }

// ── Bridge Functions ──────────────────────────────────────────────────────────

object GoogleMobileAdsFunctions {

    // ── Initialize ────────────────────────────────────────────────────────────

    class Initialize(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            mainHandler.post {
                MobileAds.initialize(activity) {
                    Log.i(TAG, "MobileAds SDK initialized")
                }
            }
            return BridgeResponse.success(mapOf("status" to "initialized"))
        }
    }

    // ── Banner ────────────────────────────────────────────────────────────────

    class ShowBanner(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val adUnitId = parameters["ad_unit_id"] as? String
                ?: return BridgeResponse.error("MISSING_PARAM", "ad_unit_id is required")

            val position = parameters["position"] as? String ?: "bottom"
            val sizeKey  = parameters["size"] as? String ?: "adaptive"

            mainHandler.post {
                val adView = AdView(activity)
                adView.adUnitId = adUnitId
                adView.setAdSize(resolveAdSize(sizeKey))

                adView.adListener = object : AdListener() {
                    override fun onAdLoaded() {
                        activity.dispatchAdEvent(
                            "NativePHP\\GoogleMobileAds\\Events\\AdLoaded",
                            mapOf("adType" to "banner", "adUnitId" to adUnitId)
                        )
                    }
                    override fun onAdFailedToLoad(error: LoadAdError) {
                        activity.dispatchAdEvent(
                            "NativePHP\\GoogleMobileAds\\Events\\AdFailedToLoad",
                            mapOf("adType" to "banner", "adUnitId" to adUnitId, "errorCode" to error.code, "errorMessage" to error.message)
                        )
                    }
                    override fun onAdOpened()     { activity.dispatchAdEvent("NativePHP\\GoogleMobileAds\\Events\\AdOpened",    mapOf("adType" to "banner")) }
                    override fun onAdClosed()     { activity.dispatchAdEvent("NativePHP\\GoogleMobileAds\\Events\\AdClosed",    mapOf("adType" to "banner")) }
                    override fun onAdImpression() { activity.dispatchAdEvent("NativePHP\\GoogleMobileAds\\Events\\AdImpression",mapOf("adType" to "banner")) }
                    override fun onAdClicked()    { activity.dispatchAdEvent("NativePHP\\GoogleMobileAds\\Events\\AdClicked",   mapOf("adType" to "banner")) }
                }

                AdHolder.bannerView?.let { old ->
                    (activity.window.decorView.rootView as? ViewGroup)?.removeView(old)
                    old.destroy()
                }

                val gravity = if (position == "top") Gravity.TOP else Gravity.BOTTOM
                val params = FrameLayout.LayoutParams(
                    FrameLayout.LayoutParams.MATCH_PARENT,
                    FrameLayout.LayoutParams.WRAP_CONTENT,
                    gravity
                )

                // Added straight to the decor view with no inset awareness, this
                // rendered flush against the physical edge — on an edge-to-edge
                // window that's on top of the status bar (position "top") or
                // behind the gesture/nav bar (position "bottom"), covering system
                // UI outright. AdMob's own placement policy prohibits ads
                // overlapping system chrome, so the relevant system-bar inset is
                // applied as a margin instead of a bare 0.
                val systemBarInsets = ViewCompat.getRootWindowInsets(activity.window.decorView)
                    ?.getInsets(WindowInsetsCompat.Type.systemBars())
                if (position == "top") {
                    params.topMargin = systemBarInsets?.top ?: 0
                } else {
                    // The system-bar inset alone only clears the gesture/nav bar,
                    // not the host app's own bottom tab bar sitting above it —
                    // this Kotlin code has no visibility into that Compose/SwiftUI
                    // layout to measure it exactly. Material 3's NavigationBar
                    // spec (80dp) is used as a reasonable estimate; a host with a
                    // taller or shorter bottom bar will need to adjust this.
                    val estimatedBottomBarHeight = (80 * activity.resources.displayMetrics.density).toInt()
                    params.bottomMargin = (systemBarInsets?.bottom ?: 0) + estimatedBottomBarHeight
                }

                (activity.window.decorView.rootView as? ViewGroup)?.addView(adView, params)
                AdHolder.bannerView = adView

                adView.loadAd(AdRequest.Builder().build())
            }

            return BridgeResponse.success(mapOf("status" to "loading"))
        }

        private fun resolveAdSize(size: String): AdSize {
            val metrics = activity.resources.displayMetrics
            val adWidth = (metrics.widthPixels / metrics.density).toInt()
            return when (size) {
                "banner"           -> AdSize.BANNER
                "large_banner"     -> AdSize.LARGE_BANNER
                "medium_rectangle" -> AdSize.MEDIUM_RECTANGLE
                else               -> AdSize.getCurrentOrientationAnchoredAdaptiveBannerAdSize(activity, adWidth)
            }
        }
    }

    class HideBanner(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            mainHandler.post {
                AdHolder.bannerView?.let { banner ->
                    (activity.window.decorView.rootView as? ViewGroup)?.removeView(banner)
                    banner.destroy()
                    AdHolder.bannerView = null
                }
            }
            return BridgeResponse.success(mapOf("status" to "hidden"))
        }
    }

    // ── Interstitial ──────────────────────────────────────────────────────────

    class LoadInterstitial(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val adUnitId = parameters["ad_unit_id"] as? String
                ?: return BridgeResponse.error("MISSING_PARAM", "ad_unit_id is required")

            mainHandler.post {
                InterstitialAd.load(activity, adUnitId, AdRequest.Builder().build(),
                    object : InterstitialAdLoadCallback() {
                        override fun onAdLoaded(ad: InterstitialAd) {
                            AdHolder.interstitial = ad
                            activity.dispatchAdEvent(
                                "NativePHP\\GoogleMobileAds\\Events\\AdLoaded",
                                mapOf("adType" to "interstitial", "adUnitId" to adUnitId)
                            )
                        }
                        override fun onAdFailedToLoad(error: LoadAdError) {
                            activity.dispatchAdEvent(
                                "NativePHP\\GoogleMobileAds\\Events\\AdFailedToLoad",
                                mapOf("adType" to "interstitial", "adUnitId" to adUnitId, "errorCode" to error.code, "errorMessage" to error.message)
                            )
                        }
                    }
                )
            }
            return BridgeResponse.success(mapOf("status" to "loading"))
        }
    }

    class ShowInterstitial(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val ad = AdHolder.interstitial
                ?: return BridgeResponse.error("NOT_LOADED", "No interstitial loaded. Call LoadInterstitial first.")

            mainHandler.post {
                AdHolder.interstitial = null
                ad.fullScreenContentCallback = fullScreenCallback(activity, "interstitial")
                ad.show(activity)
            }
            return BridgeResponse.success(mapOf("status" to "showing"))
        }
    }

    // ── Rewarded ──────────────────────────────────────────────────────────────

    class LoadRewarded(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val adUnitId = parameters["ad_unit_id"] as? String
                ?: return BridgeResponse.error("MISSING_PARAM", "ad_unit_id is required")

            mainHandler.post {
                RewardedAd.load(activity, adUnitId, AdRequest.Builder().build(),
                    object : RewardedAdLoadCallback() {
                        override fun onAdLoaded(ad: RewardedAd) {
                            AdHolder.rewarded = ad
                            activity.dispatchAdEvent(
                                "NativePHP\\GoogleMobileAds\\Events\\AdLoaded",
                                mapOf("adType" to "rewarded", "adUnitId" to adUnitId)
                            )
                        }
                        override fun onAdFailedToLoad(error: LoadAdError) {
                            activity.dispatchAdEvent(
                                "NativePHP\\GoogleMobileAds\\Events\\AdFailedToLoad",
                                mapOf("adType" to "rewarded", "adUnitId" to adUnitId, "errorCode" to error.code, "errorMessage" to error.message)
                            )
                        }
                    }
                )
            }
            return BridgeResponse.success(mapOf("status" to "loading"))
        }
    }

    class ShowRewarded(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val ad = AdHolder.rewarded
                ?: return BridgeResponse.error("NOT_LOADED", "No rewarded ad loaded. Call LoadRewarded first.")

            mainHandler.post {
                AdHolder.rewarded = null
                ad.fullScreenContentCallback = fullScreenCallback(activity, "rewarded")
                ad.show(activity, OnUserEarnedRewardListener { reward ->
                    activity.dispatchAdEvent(
                        "NativePHP\\GoogleMobileAds\\Events\\RewardEarned",
                        mapOf("rewardType" to reward.type, "rewardAmount" to reward.amount)
                    )
                })
            }
            return BridgeResponse.success(mapOf("status" to "showing"))
        }
    }

    // ── Rewarded Interstitial ─────────────────────────────────────────────────

    class LoadRewardedInterstitial(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val adUnitId = parameters["ad_unit_id"] as? String
                ?: return BridgeResponse.error("MISSING_PARAM", "ad_unit_id is required")

            mainHandler.post {
                RewardedInterstitialAd.load(activity, adUnitId, AdRequest.Builder().build(),
                    object : RewardedInterstitialAdLoadCallback() {
                        override fun onAdLoaded(ad: RewardedInterstitialAd) {
                            AdHolder.rewardedInterstitial = ad
                            activity.dispatchAdEvent(
                                "NativePHP\\GoogleMobileAds\\Events\\AdLoaded",
                                mapOf("adType" to "rewarded_interstitial", "adUnitId" to adUnitId)
                            )
                        }
                        override fun onAdFailedToLoad(error: LoadAdError) {
                            activity.dispatchAdEvent(
                                "NativePHP\\GoogleMobileAds\\Events\\AdFailedToLoad",
                                mapOf("adType" to "rewarded_interstitial", "adUnitId" to adUnitId, "errorCode" to error.code, "errorMessage" to error.message)
                            )
                        }
                    }
                )
            }
            return BridgeResponse.success(mapOf("status" to "loading"))
        }
    }

    class ShowRewardedInterstitial(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val ad = AdHolder.rewardedInterstitial
                ?: return BridgeResponse.error("NOT_LOADED", "No rewarded interstitial loaded. Call LoadRewardedInterstitial first.")

            mainHandler.post {
                AdHolder.rewardedInterstitial = null
                ad.fullScreenContentCallback = fullScreenCallback(activity, "rewarded_interstitial")
                ad.show(activity, OnUserEarnedRewardListener { reward ->
                    activity.dispatchAdEvent(
                        "NativePHP\\GoogleMobileAds\\Events\\RewardEarned",
                        mapOf("rewardType" to reward.type, "rewardAmount" to reward.amount)
                    )
                })
            }
            return BridgeResponse.success(mapOf("status" to "showing"))
        }
    }

    // ── App Open ──────────────────────────────────────────────────────────────

    class LoadAppOpen(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val adUnitId = parameters["ad_unit_id"] as? String
                ?: return BridgeResponse.error("MISSING_PARAM", "ad_unit_id is required")

            mainHandler.post {
                AppOpenAd.load(activity, adUnitId, AdRequest.Builder().build(),
                    object : AppOpenAd.AppOpenAdLoadCallback() {
                        override fun onAdLoaded(ad: AppOpenAd) {
                            AdHolder.appOpen = ad
                            activity.dispatchAdEvent(
                                "NativePHP\\GoogleMobileAds\\Events\\AdLoaded",
                                mapOf("adType" to "app_open", "adUnitId" to adUnitId)
                            )
                        }
                        override fun onAdFailedToLoad(error: LoadAdError) {
                            activity.dispatchAdEvent(
                                "NativePHP\\GoogleMobileAds\\Events\\AdFailedToLoad",
                                mapOf("adType" to "app_open", "adUnitId" to adUnitId, "errorCode" to error.code, "errorMessage" to error.message)
                            )
                        }
                    }
                )
            }
            return BridgeResponse.success(mapOf("status" to "loading"))
        }
    }

    class ShowAppOpen(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val ad = AdHolder.appOpen
                ?: return BridgeResponse.error("NOT_LOADED", "No app open ad loaded. Call LoadAppOpen first.")

            mainHandler.post {
                AdHolder.appOpen = null
                ad.fullScreenContentCallback = fullScreenCallback(activity, "app_open")
                ad.show(activity)
            }
            return BridgeResponse.success(mapOf("status" to "showing"))
        }
    }
}
